Examples visual

tias opened the pull request:
Reworked visual examples: as notebooks, and MUCH simpler visualisations with the usual visualisation tools (pandas, plotly)

@IgnaceBleukx can you do a code review? I tried to follow the style that we discussed for the csplib examples. Also, a somewhat different way of checking the result of model.solve() while still being able to work at the 'top level' of a cell...

Do you know of other examples that have a visualisation that we could add?

Maybe VRP/TSP? we have 'some' visualisation in advanced/VRP by learning...ipynb

IgnaceBleukx commented on examples/visual/car_sequencing.ipynb:
```diff
@@ -0,0 +1,302 @@
+    "    requires = cpm_array(requires)  # For element constraint\n",
+    "\n",
+    "    # Decision variables\n",
+    "    slots = intvar(1, nCarConfigs, shape=nSlots, name=\"slots\")  # Sequence of different car types\n",
```
Writing it 0-indexed might be a little clearer in the constraint themselves? But 1-indexed seems the way it is done most of the time indeed....

IgnaceBleukx commented on examples/visual/car_sequencing.ipynb:
```diff
@@ -0,0 +1,302 @@
+    "    # Check that no more than \"at most\" car properties are used per \"per_slots\" slots\n",
+    "    for o in range(nOptions):\n",
+    "        for s in range(nSlots - per_slots[o]):\n",
+    "            slotrange = range(s, s + per_slots[o])\n",
```
In the csp-lib version I used the `rolling_window` function from numpy. For me it's more clear using that one but this is fine too

IgnaceBleukx commented on examples/visual/packing_rectangles.ipynb:
```diff
@@ -0,0 +1,197 @@
+    "          pos_y + heights <= total_y]\n",
+    "\n",
+    "    # No-overlap: every item has to be fully above, below or next to every other item\n",
+    "    for i in range(n):\n",
```
Why not use the `all_pairs` helper function?

IgnaceBleukx reviewed (commented):
Diff of scheduling.ipynb was too large for github to review:
- For the precedences, I like to use `start[:,-1] <= end[:,1:]`
- Same for the optional constraint of job B
- I would also use snake_case in this file

IgnaceBleukx commented:
We could also add a nonogram example? Depending on the picture, it could be really nice

IgnaceBleukx commented:
Other problems on the wishlist:
- RCPSP
- Nonogram (see examples/csplib/prob012_nonogram.py) which can be made cleaner after merging #121

tias commented on examples/visual/car_sequencing.ipynb:
```diff
@@ -0,0 +1,302 @@
+    "    df = pd.DataFrame(setup).astype(str).replace('True', 'X').replace('False', '')\n",
+    "    df['Config (slots)'] = [f\"Options {l}\" for l in slots]  # insert as first column\n",
+    "    df = df.set_index('Config (slots)')\n",
+    "    return df"
```
Yes, do you have an example/snippet?

tias commented:
Addressed the earlier comments.

Added visual tables for nqueens and sudoku, pretty cute.

Moved all visual examples to the top-level examples folder, and made a README that lists the notebooks and what they are examples of.

Ideally there would be colab/binder/github links for each of these, but I did not have the patience for that.


I will merge this branch by the end of the week so the CSAI students can see/use them. Any comments/additions welcome till then.
